saw-core-coq/src/Verifier/SAW/Translation/Coq/Term.hs

```diff
@@ -443,85 +512,65 @@ mkDefinition name (Coq.Lambda bs t) (Coq.Pi bs' tp)
--- and zero or more additional 'Ident's and 'Type's representing additonal
--- implicit typeclass arguments, added if the given type is @isort@, etc.
-translateBinder ::
-  TermTranslationMonad m =>
-  LocalName ->
-  Term ->
-  m (Coq.Ident,Coq.Type,[(Coq.Ident,Coq.Type)])
-translateBinder n ty@(asPiList -> (args, asSortWithFlags -> mb_sort)) =
+-- | The result of translating a SAW core variable binding to Coq, including the
+-- Coq identifier for the variable, the Coq translation of its type, and 0 or
+-- more implicit Coq arguments that apply to the variable
+data BindTrans = BindTrans { bindTransIdent :: Coq.Ident,
+                             bindTransType :: Coq.Type,
+                             bindTransImps :: [(Coq.Ident,Coq.Type)] }
+
+-- | Convert a 'BindTrans' to a list of Coq term-level binders
+bindTransToBinder :: BindTrans -> [Coq.Binder]
+bindTransToBinder (BindTrans {..}) =
+  Coq.Binder bindTransIdent (Just bindTransType) :
+  map (\(n,ty) -> Coq.ImplicitBinder n (Just ty)) bindTransImps
+
+-- | Convert a 'BindTrans' to a list of Coq type-level pi-abstraction binders
+bindTransToPiBinder :: BindTrans -> [Coq.PiBinder]
+bindTransToPiBinder (BindTrans { bindTransImps = [], .. })
+  | bindTransIdent == "_" = [Coq.PiBinder Nothing bindTransType]
+bindTransToPiBinder (BindTrans { bindTransImps = [], .. }) =
+  [Coq.PiBinder (Just bindTransIdent) bindTransType]
+bindTransToPiBinder (BindTrans{..}) =
+  Coq.PiBinder (Just bindTransIdent) bindTransType :
+  map (\(n,ty) -> Coq.PiImplicitBinder (Just n) ty) bindTransImps
+
+-- | Given a 'LocalName' and its type (as a 'Term'), translate the 'LocalName'
+-- to a Coq identifier, translate the type to a Coq term, and generate zero or
+-- more additional 'Ident's and 'Type's representing additonal implicit
+-- typeclass arguments, added if the given type is @isort@, etc. Pass all of
+-- this information to the supplied computation, in which the SAW core variable
+-- is bound to its Coq identifier.
+translateBinder :: TermTranslationMonad m => LocalName -> Term ->
+                   (BindTrans -> m a) -> m a
+translateBinder n ty@(asPiList -> (args, asSortWithFlags -> mb_sort)) f =
   do ty' <- translateTerm ty
-     n' <- freshenAndBindName n
      let flagValues = sortFlagsToList $ maybe noFlags snd mb_sort
          flagLocalNames = [("Inh", "SAWCoreScaffolding.Inhabited"),
                            ("QT", "QuantType")]
-     nhs <- forM (zip flagValues flagLocalNames) $ \(fi,(prefix,tc)) ->
-       if not fi then return []
-       else do nhty <- translateImplicitHyp (Coq.Var tc) args (Coq.Var n')
-               nh <- translateLocalIdent (prefix <> "_" <> n)
-               return [(nh,nhty)]
-     return (n',ty',concat nhs)
-
--- | Given a typeclass (as a 'Term'), a list of 'LocalName's and their
+     withSAWVar n $ \n' ->
+       helper n' (zip flagValues flagLocalNames) (\imps ->
+                                                   f $ BindTrans n' ty' imps)
```

Comment:
This is not new to this PR, but I'm a little spooked by the seemingly non-local assumption that whatever code is setting `mb_sort` is in sync with this here list of `flagLocalNames`.

Reply:
Yeah, I'm confused by that part as well